### DISCUSSION TITLE
Questions/Ideas about the Jets and the Robust Cost

### DISCUSSION THREAD

**Discussion by yimuw** [ceres-solver/ceres-solver#648] [Ideas]
Hi!

I am a huge fan of Ceres! Thanks for developing such a great library!

I tried to implement Ceres from scratch for fun and wrote a study note for it. I got a few questions during my study and implementation.

1. Where using Jets? My feeling is Jets are a pre-allocated variable for tracking the derivative w.r.t **ALL** inputs. Why not just tracking the derivative w.r.t existing inputs instead of the derivative w.r.t all inputs? [detail discussion in my study note](https://wang-yimu.com/ceres-solver/#ceres-jet) and [a POC of tracking the existing variables](https://wang-yimu.com/ceres-solver/#dictionary-impl)

2. The [Robust Loss derivation](http://ceres-solver.org/nnls_modeling.html#theory) from the Ceres Doc (potentially) could be simplified with possible performance again.  [An alternative derivation for the robust loss](https://wang-yimu.com/ceres-solver/#robust-loss)

3. A few not so important questions/requests. (a) adding API for adding quadratic cost; (b) Why using trust-region instead of line search? (c) Adding local parameters for SO(3). [link](https://wang-yimu.com/ceres-solver/#questions)

Let me know if I missing anything.

Thanks again for such a great library!

**Comment by sandwichmaker** [comment]
Hi @yimuw,
sorry for the delayed reply, its the holidays. Let me try and answer your questions.

1. Where using Jets? My feeling is Jets are a pre-allocated variable for tracking the derivative w.r.t ALL inputs. Why not just tracking the derivative w.r.t existing inputs instead of the derivative w.r.t all inputs? detail discussion in my study note and a POC of tracking the existing variables

As you observe in your note, this is the difference between forward and reverse mode differentiation. Jets are doing forward mode which for small dense residual blocks is a good idea. But as the size of the jet increases their computational complexity also goes up and backprop/reverse mode is better. If you come up with a more efficient backward mode implementation I'd be curious to look at it.

2. The Robust Loss derivation from the Ceres Doc (potentially) could be simplified with possible performance again. An alternative derivation for the robust loss

Your derivation is equivalent to assuming that \rho'' = 0.

3. A few not so important questions/requests. (a) adding API for adding quadratic cost; (b) Why using trust-region instead of line search? (c) Adding local parameters for SO(3). link

I do not understand what you mean by API for quadratic cost.

Generally speaking, Trust region methods like Levenberg-Marquardt have worked better than line search methods for non-linear least squares problems. We have l-bfgs with line search in ceres, you can experiment with it if you like. But we do not have a full newton based line search.

We have local parameterization for quaternions. It is straightforward to add a local parameterization for 3x3 matrices in SO(3).
